pkgs/development/tools/rust/rustup/default.nix

```diff
@@ -57,6 +57,16 @@ rustPlatform.buildRustPackage rec {
 
   doCheck = !stdenv.isAarch64 && !stdenv.isDarwin;
 
+  # Currently rustup requires 1GB per compile thread, which exceed's some
+  # contributors memory. Therefore we limit parallism to 2/3 of the available
+  # cores. This is a bit adhoc based on current common hardware configuration
+  # and might need to be adjusted in future.
+  preBuild = ''
+    if [[ "$NIX_BUILD_CORES" -ge "$(nproc)" ]]; then
+      (( NIX_BUILD_CORES = $(nproc) * 66 / 100 ))
```

Comment:
An alternative might be to use memory but this seems to be linux specific (i.e. the free program) whereas `nproc` is part of coreutils. Also makes the code simpler.